src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+  private final Set<String> defaultRoles;
+  private final String authorizationMode;
+  private final String authorizationUrl;
+
+  public TeamProjectBasedAuthorizingFilter(
+          PermissionManager authorizer,
+          Set<String> defaultRoles,
+          String authorizationMode,
+          String authorizationUrl) {
+    this.authorizer = authorizer;
+    this.defaultRoles = defaultRoles;
+    this.authorizationMode = authorizationMode;
+    this.authorizationUrl = authorizationUrl;
+    logger.debug("AUTHORIZATION_MODE in TeamProjectBasedAuthorizingFilter constructor == '{}'", this.authorizationMode);
+    logger.debug("AUTHORIZATION_URL in TeamProjectBasedAuthorizingFilter constructor == '{}'", this.authorizationUrl);
+  }
+
+  @Override
+  protected boolean preHandle(ServletRequest request, ServletResponse response) throws Exception {
+
+    try {
+        logger.debug("preHandle in TeamProjectBasedAuthorizingFilter == '{}'", this.authorizationMode);
+        if (this.authorizationMode.equals("teamproject") && SecurityUtils.getSubject().isAuthenticated()) {
+          // in case of "teamproject" mode, we want all roles to be reset always, and
+          // set to only the one requested/found in the request parameters (following lines below):
+          String login = this.authorizer.getCurrentUser().getLogin();
+          boolean foundValidTeamProject = extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(this, login, request, response);
+          if (!foundValidTeamProject) {
+            return false;
+          }
+        }
+
+      } catch (Exception e) {
+        WebUtils.toHttp(response).setHeader("x-auth-error", e.getMessage());
+        throw new Exception(e);
+      }
+
+    return true;
+  }
+
+  /**
+   * Tries to extract "team project" role found in the request.
+   * If necessary, assigns the newly found "team project" role to the user,
+   * after validating with a Gen3 authorization service if the user
+   * has been granted access to this team.
+   * 
+   * @param self: reference to "this". Just a workaround to allow the method to be
+   * globally synchronized by making it static.
+   * @param login: the user login
+   * @param request: the request passing through this filter
+   * @param response: response object that can be used to write error status and
+   * error message if needed.
+   * 
+   * @return: Returns true if the "team project" was found and passed validation
+   * and DB update to user roles did not fail.
+   * 
+   * @throws IOException
+   * @throws Exception
+   */
+  private static boolean extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(
+      TeamProjectBasedAuthorizingFilter self,
+      String login, ServletRequest request,
+      ServletResponse response)  throws IOException, Exception {
+
+    // synchronize on login to avoid race conditions (especially on DB updates) if
+    // this filter receives parallel requests for any reason:
+    synchronized (login.intern()) {
```

Comment:
Question: does this work for both POST and GET? I think that the examples I have seen have the team project in the GET only but wonder if that could be an issue if folks can see and decide to copy URLS? For example, other people (hackers) seeing the url copy will learn about what teams exist.

Reply:
A malicious user can add any teamproject to the URL, there is no stopping that. We do however handle this case by validating whether the user has the right teamproject authorization (in Gen3 Arborist in this case).

Reply:
Ok, but do you all recommend the use POST or the header? if so, maybe we could get rid of the GET option? That might lower the chance of team project identifiers being copied on places where bots and hackers are sniffing.

Reply:
marking resolved, as discussed today.